tstrait/genetic.py

```diff
@@ -0,0 +1,207 @@
+import numba
+import numpy as np
+import pandas as pd
+import tskit
+
+from .base import _check_instance, _check_dataframe, _check_non_decreasing  # noreorder
+
+
+@numba.njit
+def _traversal_genotype(
+    nodes_individual,
+    left_child_array,
+    right_sib_array,
+    stack,
+    has_mutation,
+    num_individuals,
+    num_nodes,
+):  # pragma: no cover
+    """
+    Numba to speed up the tree traversal algorithm to determine the genotype of
+    individuals.
+    Stack has to be Typed List in numba to use numba.
+    """
+
+    genotype = np.zeros(num_individuals)
+    while len(stack) > 0:
+        parent_node_id = stack.pop()
+        if parent_node_id == num_nodes:
+            individual_id = -1
+        else:
+            individual_id = nodes_individual[parent_node_id]
+        if individual_id > -1:
+            genotype[individual_id] += 1
+        child_node_id = left_child_array[parent_node_id]
+        while child_node_id != -1:
+            if not has_mutation[child_node_id]:
+                stack.append(child_node_id)
+            child_node_id = right_sib_array[child_node_id]
+
+    return genotype
+
+
+class _GeneticValue:
+    """GeneticValue class to compute genetic values of individuals.
+
+    Parameters
+    ----------
+    ts : tskit.TreeSequence
+        Tree sequence data with mutation
+    trait_df : pandas.DataFrame
+        Dataframe that includes causal site ID, causal state, simulated effect
+        size, and trait ID.
+    alpha : float
+        Parameter that determines the relative weight on rarer variants.
+    """
+
+    def __init__(self, ts, trait_df):
+        self.trait_df = trait_df[["site_id", "effect_size", "trait_id", "causal_state"]]
+        self.ts = ts
+
+    def _individual_genotype(self, tree, site, causal_state, num_nodes):
```

Comment:
There's isn't really much point in counting the causal states, and we may as well just accumulate the actual values here and save the multiply later. Something like
```
def _individual_genetic_values(self, tree, site, causal_state, effect_size):
     has_mutation = np.zeros(self.ts.num_nodes + 1, dtype=bool)